toolchain/base/int_store.h

```diff
@@ -0,0 +1,186 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_TOOLCHAIN_BASE_INT_STORE_H_
+#define CARBON_TOOLCHAIN_BASE_INT_STORE_H_
+
+#include "common/check.h"
+#include "llvm/ADT/APFloat.h"
+#include "llvm/ADT/APInt.h"
+#include "llvm/ADT/SmallVector.h"
+#include "toolchain/base/mem_usage.h"
+#include "toolchain/base/value_ids.h"
+#include "toolchain/base/value_store.h"
+#include "toolchain/base/yaml.h"
+
+namespace Carbon {
+
+// Forward declare a testing peer so we can friend it.
+namespace Testing {
+struct IntStoreTestPeer;
+}  // namespace Testing
+
+// A canonicalizing value store with deep optimizations for integers.
+//
+// This stores integers as abstract, signed mathematical integers. The bit width
+// of specific `APInt` values, either as inputs or outputs, is disregarded for
+// the purpose of canonicalization and the returned integer may use a very
+// different bit width `APInt` than was used when adding. There are also
+// optimized paths for adding integer values representable using native integer
+// types.
+//
+// Because the integers in the store are canonicalized without a specific bit
+// width there are helper functions to coerce them to a specific desired bit
+// width for use.
+//
+// This leverages a significant optimization for small integer values -- rather
+// than canonicalizing and making them unique in a `ValueStore`, they are
+// directly embedded in the `IntId` itself. Only larger integers are stored in
+// an array of `APInt` values and represented as an index in the ID.
+class IntStore {
+ public:
+  // Adds an integer value representable in a host `int64_t` to the store.
+  // Especially useful when the integer is computed without an `APInt` in the
+  // first place.
+  //
+  // This only accepts a signed `int64_t` and uses the mathematical signed
+  // integer value of it as the added integer value.
+  //
+  // Returns the ID corresponding to this integer value, storing an `APInt` if
+  // necessary to represent it.
+  auto Add(int64_t value) -> IntId {
+    // First try directly making this into an ID.
+    if (IntId id = IntId::TryMakeValue(value); id.is_valid()) [[likely]] {
+      return id;
+    }
+
+    // Fallback for larger values.
+    return AddLarge(value);
+  }
+
+  // Stores a canonical copy of a signed value and returns its ID.
```

Comment:
Isn't this only conditionally going to store a copy? Maybe similar to the above:

```suggestion
  // Returns the ID corresponding to this integer value, storing an `APInt` if
  // necessary to represent it.
```